File: problems/algorithms/0050/Rust/solution_bitwise.rs
impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        if n == 0 {
            return 1.0;
        }
        let mut pow = 1.0;
        let abs_exp = n.abs() as u32;
        let mut num: u32 = 1 << 31;
        for i in 0..32 {
            let is_bit_one = (abs_exp & num != 0);
            pow *= pow;
            if is_bit_one {
                pow *= x;
            }
            num = num >> 1;
        }
        if n < 0 {
            return 1.0 / pow;
        }
        pow
    }
}

File: problems/algorithms/0050/Cpp/solution_bitwise.cpp
class Solution
{
public:
	double myPow(double x, int n)
	{
		if (n == 0.0)
			return 1.0;
		if (n == 1.0 || x == 1.0)
			return x;
		double pow = 1.0;
		long abs_exponent = (long)n;
		if (abs_exponent < 0)
			abs_exponent *= -1;
		unsigned int num = 1 << 31;
		for (int i = 0; i < 32; i++)
		{
			int is_bit_one = ((abs_exponent & num) != 0);
			pow *= pow;
			if (is_bit_one)
			{
				pow *= x;
			}
			num = num >> 1;
		}
		if (n < 0)
			return 1.0 / pow;
		return pow;
	}
};